groth16/src/lib.rs

```diff
@@ -48,7 +51,7 @@ mod test;
 pub use self::{generator::*, prover::*, verifier::*};
 
 /// A proof in the Groth16 SNARK.
-#[derive(Clone)]
+#[derive(Clone, CanonicalSerialize, CanonicalDeserialize)]
```

Comment:
For these impls to work, you would probably need to either make `PairingEngine: CanonicalSerialize + CanonicalDeserialize`, or you would need to adjust the derive macros to take in a separate `bounds` parameter like in [`derivative`](https://mcarton.github.io/rust-derivative/Debug.html#custom-bound) or in [`serde`](https://serde.rs/attr-bound.html).

Reply:
(The latter is preferable, but is also more work, so it's up to you)

Reply:
No, I don't think you need this. The only requirement for the derive macro to work is that all fields within the struct implement CanonicalSerialize/CanonicalDeserialize (and it will fail during compilation if this requirement is not given).

`E::G1Affine` and `E::G2Affine` both have the trait bound `AffineCurve`, which requires `CanonicalSerialize`/`CanonicalDeserialize` to be implemented. Thus, the derivation should work.

For reference, here's the code the compiler currently produces for exactly this derivation here:
```
impl <E: PairingEngine> CanonicalSerialize for Proof<E> {
    #[allow(unused_mut, unused_variables)]
    fn serialize<W: ::algebra_core::io::Write>(&self, writer: &mut W)
     -> Result<(), ::algebra_core::SerializationError> {
        CanonicalSerialize::serialize(&self.a, writer)?;
        CanonicalSerialize::serialize(&self.b, writer)?;
        CanonicalSerialize::serialize(&self.c, writer)?;
        Ok(())
    }
    #[allow(unused_mut, unused_variables)]
    fn serialized_size(&self) -> usize {
        let mut size = 0;
        size += CanonicalSerialize::serialized_size(&self.a);
        size += CanonicalSerialize::serialized_size(&self.b);
        size += CanonicalSerialize::serialized_size(&self.c);
        size
    }
}
impl <E: PairingEngine> CanonicalDeserialize for Proof<E> {
    #[allow(unused_mut, unused_variables)]
    fn deserialize<R: ::algebra_core::io::Read>(reader: &mut R)
     -> Result<Self, ::algebra_core::SerializationError> {
        {
            Ok(Proof{a:
                         ::algebra_core::CanonicalDeserialize::deserialize(reader)?,
                     b:
                         ::algebra_core::CanonicalDeserialize::deserialize(reader)?,
                     c:
                         ::algebra_core::CanonicalDeserialize::deserialize(reader)?,})
        }
    }
}
```

Reply:
Interesting, the default derive macros from the `std` library add trait bounds to each type parameter.